Feature/newsletter

JesusArtiles opened the pull request:
In this branch, we manage the creation and mailing of an excel with the volunteers subscribed to the newsletter.

Commit:
Add: Prepared new endpoint for changing status of newsletter subscription(not functional until merge).

Commit:
Add: new endpoint for downloading excel with subscribed volunteers.

Commit:
Refactor: refactor ManageNewsletterExcelAction
Add: new method in JpaVolunteerRepository

juanmaMacGyverCode committed:
* MERGE: Merged new features from develop

JeremyTrujillo committed:
* ADD: Added field in database

JeremyTrujillo committed:
* ADD: Added field to Profile
* REFACTOR: Moved newsletter methods to corresponding controller
* REFACTOR: Refactored ChangeStatusNewsletterSubscriptionAction
* ADD: Added test case

JeremyTrujillo committed:
* FIX: Response Status
* ADD: Added test case
* ADD: Added method in TestData

Commit:
* Add: New template and support in email service to send newsletter email.
* Add: support in remote Storage service to upload the newsletter excel file.
* Add: Newsletter Action now storages the excel and sends the email to reviser.

JeremyTrujillo committed:
* ADD: Added new email template
* ADD: Added support for emptyNewsletter
* ADD: Added test cases
* FIX: Fixed error in TestData

JeremyTrujillo committed:
* FIX: Fixed SonarLint issues

juanmaMacGyverCode commented on backend/src/main/java/com/huellapositiva/application/controller/NewsletterController.java:
```diff
@@ -0,0 +1,79 @@
+    @Autowired
+    private final ChangeStatusNewsletterSubscriptionAction changeStatusNewsletterSubscriptionAction;
+
+    @Operation(
+            summary = "Change status of subscription to newsletter",
+            description = "Changes the status of the subscribed parameter on the specified volunteer",
+            tags = "newsletter"
+    )
+    @ApiResponses(
+            value = {
+                    @ApiResponse(
+                            responseCode = "200",
+                            description = "Ok, status of subscribed field changed successfully"
+                    ),
+                    @ApiResponse(
+                            responseCode = "500",
+                            description = "Internal server error, could not fetch the user data due to a connectivity issue."
+                    )
+            }
+    )
```
Check it with the new swagger. (VolunteerApiController)

JesusArtiles replied:
Solved

juanmaMacGyverCode commented on backend/src/main/java/com/huellapositiva/application/controller/NewsletterController.java:
```diff
@@ -0,0 +1,79 @@
+    @RolesAllowed("VOLUNTEER")
+    @ResponseBody
+    @ResponseStatus(HttpStatus.OK)
+    public void changeStatusNewsletterSubscription(@RequestBody Boolean subscribed,
+                                                   @AuthenticationPrincipal String volunteerEmail) {
```
Add notation @Parameter(hidden = true) in everything @AuthenticationPrincipal

JesusArtiles replied:
Solved

juanmaMacGyverCode commented on backend/src/main/java/com/huellapositiva/application/controller/NewsletterController.java:
```diff
@@ -0,0 +1,79 @@
+        changeStatusNewsletterSubscriptionAction.execute(subscribed, volunteerEmail);
+    }
+
+    @Operation(
+            summary = "Download newsletter excel",
+            description = "Prepare an excel with the volunteers subscribed and send it to reviser",
+            tags = "newsletter"
+    )
+    @ApiResponses(
+            value = {
+                    @ApiResponse(
+                            responseCode = "204",
+                            description = "No content, email has been verified"
+                    ),
+                    @ApiResponse(
+                            responseCode = "500",
+                            description = "Internal server error, could not fetch the user data due to a connectivity issue."
+                    )
+            }
+    )
```
Same

JesusArtiles replied:
Solved

juanmaMacGyverCode commented on backend/src/main/java/com/huellapositiva/application/controller/NewsletterController.java:
```diff
@@ -0,0 +1,79 @@
+                    )
+            }
+    )
+    @GetMapping("/download")
+    @RolesAllowed("REVISER")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void downloadExcel(@AuthenticationPrincipal String reviserEmail) throws IOException {
```
download isnt the purpose, it's "sendLinkExcelInEmail" for example. Revise the URI, isnt the purpose.

JesusArtiles replied:
Solved

juanmaMacGyverCode commented on backend/src/main/java/com/huellapositiva/domain/actions/ManageNewsletterExcelAction.java:
```diff
@@ -0,0 +1,69 @@
+    private static final String ROOT = "./newsletterEmails.xlsx";
+
+    public void execute(String email) throws IOException {
+        List<JpaVolunteer> subscribedVolunteers = jpaVolunteerRepository.findSubscribedVolunteers();
+        if (subscribedVolunteers.isEmpty()){
+            communicationService.sendEmptyNewsletter(EmailAddress.from(email));
+        } else {
```
Might to send an Exception and not an email?

JesusArtiles replied:
Solved

juanmaMacGyverCode commented on backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaVolunteerRepository.java:
```diff
@@ -39,4 +40,7 @@
     @Query("UPDATE JpaVolunteer p SET p.location = :location WHERE p.id = :id")
     Integer updateLocation(@Param("id") String id, @Param("location") JpaLocation location);
+
+    @Query("FROM JpaVolunteer v LEFT JOIN FETCH v.credential c WHERE v.profile.subscribed = true")
+    List<JpaVolunteer> findSubscribedVolunteers();
```
For Hibernate is better to use the complete QUERY.

JesusArtiles replied:
Solved

juanmaMacGyverCode commented on backend/src/test/java/com/huellapositiva/api/NewsletterControllerShould.java:
```diff
@@ -0,0 +1,95 @@
+        testData.resetData();
+    }
+
+    @Test
+    void return_200_when_status_changed_successfully() throws Exception {
+        testData.createVolunteerWithProfile(DEFAULT_EMAIL, DEFAULT_PASSWORD);
```
What status?

JesusArtiles replied:
Ok

juanmaMacGyverCode commented on backend/src/test/java/com/huellapositiva/api/NewsletterControllerShould.java:
```diff
@@ -0,0 +1,95 @@
+        assertThat(volunteer.getProfile().isSubscribed()).isFalse();
+    }
+
+    @Test
+    void return_204_when_excel_downloaded_successfully() throws Exception {
+        JpaVolunteer volunteer = testData.createSubscribedVolunteer(DEFAULT_EMAIL, DEFAULT_PASSWORD);
```
Send an email, not download.

JesusArtiles replied:
Ok

juanmaMacGyverCode commented on backend/src/test/java/com/huellapositiva/integration/ManageNewsletterExcelActionShould.java:
```diff
@@ -0,0 +1,72 @@
+        testData.resetData();
+    }
+
+    @Test
+    void send_an_empty_newsletter_email_when_there_are_no_subscribed_volunteers() throws IOException {
+        // GIVEN
+        ManageNewsletterExcelAction manageNewsletterExcelAction = new ManageNewsletterExcelAction(
+                jpaVolunteerRepository, remoteStorageService, emailCommunicationService);
+
+        // WHEN
+        manageNewsletterExcelAction.execute(DEFAULT_EMAIL);
+
+        // THEN
+        verify(emailCommunicationService).sendEmptyNewsletter(EmailAddress.from(DEFAULT_EMAIL));
+    }
+
+    @Test
+    void send_a_newsletter_email_when_there_are_subscribed_volunteers() throws IOException {
+        testData.createSubscribedVolunteer(DEFAULT_EMAIL, DEFAULT_PASSWORD);
+        // GIVEN
+        ManageNewsletterExcelAction manageNewsletterExcelAction = new ManageNewsletterExcelAction(
+                jpaVolunteerRepository, remoteStorageService, emailCommunicationService);
+
+        // WHEN
+        manageNewsletterExcelAction.execute(DEFAULT_EMAIL);
+
+        // THEN
+        verify(emailCommunicationService).sendNewsletter(EmailAddress.from(DEFAULT_EMAIL),null);
+    }
+}
```
Unit test or integration test?

JesusArtiles replied:
Ok I changed it from integration to unit

Commit:
Add: added @Parameter(hidden = true) teg in NewsletterController

Commit:
Refactor: Change name of the method that sands email with excel

Commit:
Refactor: Change name of test methods of newsletter

ibaiul commented on backend/src/test/java/com/huellapositiva/util/TestData.java:
```diff
@@ -397,9 +402,9 @@ public String registerESALandPublishedProposalObject() throws ParseException {
-                .startingProposalDate(ProposalDate.createStartingProposalDate("20-01-2021"))
-                .closingProposalDate(ProposalDate.createClosingProposalDate("24-01-2021"))
-                .startingVolunteeringDate(ProposalDate.createStartingVolunteeringDate("25-01-2021"))
+                .startingProposalDate(ProposalDate.createStartingProposalDate("20-11-2021"))
+                .closingProposalDate(ProposalDate.createClosingProposalDate("24-11-2021"))
+                .startingVolunteeringDate(ProposalDate.createStartingVolunteeringDate("25-11-2021"))
```
This will just delay the issue to November. It would be a better approach to user relative times instead.
Instant.now().minusDays(1) and so to guarantee that the test will not break depending on which day you run it.

ALREADY FIXED in the develop branch, just bring the changes to your branch

JesusArtiles replied:
Solved

ibaiul commented on backend/src/test/java/com/huellapositiva/integration/ManageNewsletterExcelActionShould.java:
```diff
@@ -0,0 +1,72 @@
+import static org.mockito.Mockito.verify;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@ExtendWith(MockitoExtension.class)
```
Redundant? SpringBootTest annotation probably does that already, can verify though

JesusArtiles replied:
Yes, I removed the extendwith annotation

Commit:
Added: New Exception when there are no volunteers subscribed to newsletter
Removed: Funcion to send email when there are no volunteers subscribed to newsletter

Commit:
Refactor: Changed newsletter integration test to unit test

Commit:
Refactor: Changed repository method query

Commit:
Refactor: Changed redundant annotation in tests

Commit:
REFACTOR: change IOException for IllegalStateException in ManageNewsletterExcelAction

Commit:
ADD: Changes from develop
ADD: Add information to swagger

ibaiul commented on backend/src/main/java/com/huellapositiva/application/controller/NewsletterController.java:
```diff
@@ -0,0 +1,105 @@
+                    )
+            }
+    )
+    @PostMapping("/changeStatusNewsletterSubscription")
```
El newsletter no es un resource tal y como lo pueden ser un volunteer, contact_person, ...

El cambiar los settings de un voluntario deberia ser algo como POST /api/v1/volunteer/profile/newsletter

JeremyTrujillo replied:
Ya lo hemos trasladado a VolunteerApiController.

ibaiul commented on backend/src/main/java/com/huellapositiva/application/controller/NewsletterController.java:
```diff
@@ -0,0 +1,105 @@
+    @GetMapping("/getNewsletterExcel")
+    @RolesAllowed("REVISER")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void sendExcelLinkEmail(@Parameter(hidden = true)@AuthenticationPrincipal String reviserEmail) throws IOException {
```
Es una MUY MUY MUY MUY MUY mala idea tener un endpoint accesible al publico que devuelve informacion privada de los usuarios. Esta aplicacion es de codigo libre y cualquiera puede leerla, si alguien encontrase un bug en la version de Spring, Jackson, ... que usamos y lo explotase para obtener esos email estariamos jodidos. Habria que informar a las autoridades de que ha ocurrido una violacion de la GDPR y hemos sufrido un acceso no autorizado a informacion protegida de nuestros usuarios. Es obligatorio por ley incluir esas clausulas en los terminos de uso de cualquier servicio que guarde datos protegios por las leyes europeas sobre los ciudadanos europeos. La asociacion se podria enfrentar a sanciones por no haber protegido esa informacion debidamente

Una alternativa es enviar un email al revisor que ha pedido eso con el archivo excel adjuntado.

UPDATE: Me he colado, la funcionalidad esta bien. Sorry
Dejo la explicacion para entender porque tenemos que enviarlo a traves de un email.

ibaiul commented on backend/src/main/java/com/huellapositiva/application/controller/NewsletterController.java:
```diff
@@ -0,0 +1,105 @@
+    @RolesAllowed("VOLUNTEER")
+    @ResponseBody
+    @ResponseStatus(HttpStatus.OK)
+    public void changeStatusNewsletterSubscription(@RequestBody Boolean subscribed,
```
Debemos usar objetos que se puedan representar en JSON para los request/response body. Somos una REST API en JSON.

`
@Value
public class UpdateNewsletterSubscriptionResponseDto {
    boolean enabled;
}`

Si usamos objetos representables en JSON y en un futuro incluimos o cambiamos propiedades del DTO, podemos hacer que nuestra API sea "backwards compatible", cosa que es imprescindible cuando disenamos APIs del tipo que sean.

Si usamos un String o un Boolean, no podemos versionar nuestras APIs de manera adecuada ya que romperiamos la compatibilidad con clientes existentes.

JeremyTrujillo replied:
Creamos un Dto para el request body y el método responde con un NO_CONTENT.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/actions/ManageNewsletterExcelAction.java:
```diff
@@ -0,0 +1,69 @@
+    public void execute(String email) throws IOException {
+        List<JpaVolunteer> subscribedVolunteers = jpaVolunteerRepository.findSubscribedVolunteers();
+        if (subscribedVolunteers.isEmpty()){
+            throw new IllegalStateException("Can not find any volunteers subscribed to newsletter");
```
Esto no es ningun error, si no tenemos usuarios subscritos al newsletter pues enviamos un email al revisor diciendole eso.

JeremyTrujillo replied:
Ahora enviamos un email específico cuando no hay voluntarios suscritos.

ibaiul commented on backend/src/main/java/com/huellapositiva/domain/service/RemoteStorageService.java:
```diff
@@ -64,4 +65,11 @@ public URL uploadVolunteerPhoto(MultipartFile photo, String volunteerId) throws
+        String extension = ".xlsx";
+        String destinationFileName = UUID.randomUUID() + extension;
+        String volunteerExcelRootKey = "newsletter/";
+        return storageService.upload(volunteerExcelRootKey + destinationFileName, excel, "application/vnd.ms-excel");
```
Este archivo no puede subirse a S3 con permisos publicos debido a su contenido. Tenemos que subirlo con permisos privados y crear los correspondientes roles/permisos en AWS para que solo ciertos usuarios puedan descargarlo.

Alternativamente, podriamos encriptar el archivo antes de subirlo a S3 o encriptarlo directamente ahi y mandarle la contrasena del archivo en el correo al reviser.

ibaiul commented on backend/src/main/java/com/huellapositiva/infrastructure/orm/entities/JpaProfile.java:
```diff
@@ -55,4 +55,7 @@
         private String additionalInformation;
+
+        @Column(name = "subscribed")
+        private boolean subscribed;
```
ese nombre no me dice mucho sobre su significado. Que tal si lo llamamos -> boolean newsletter?

JeremyTrujillo replied:
Ya lo renombramos.

JeremyTrujillo committed:
* REFACTOR: Moved method.
* ADD: Added empty newsletter email.

JeremyTrujillo committed:
* REFACTOR: Renamed field.

JeremyTrujillo committed:
* ADD: Added Dto.

ibaiul committed:
Review comments

ibaiul committed:
Merge branch 'develop' into feature/Newsletter
